syntax = "proto3";

message MetaData {
    map<string, string> info = 1;
}

message Source {
    string name = 1;
    repeated string processors = 2;
}

message Module {
    enum ModuleType {
        unknown = 0;
        scaler = 1;
        resampler = 2;
        vp9 = 10;
        h264 = 11;
        h265 = 12;
        h264_svc = 13;
        pcma = 20;
        pcmu = 21;
        pcm = 22;
        aac = 23;
        mp3 = 24;
    }
    ModuleType type = 1;
    int32 width = 2;
    int32 height = 3;
    int32 sample_rate = 4;
    int32 channel_num = 5;
    int32 bitrate = 6;
    int32 key_frame_interval = 7;
}

message Processor{
    enum ProcessMediaType {
        unknown = 0;
        audio = 1;
        video = 2;
        data = 3;
    }
    string name = 1;
    ProcessMediaType type = 2;
    repeated Module modules = 3;
}

message Drain {
    string name = 1;
    enum DrainType {
        unknown = 0;
        mp4 = 1;
        wav = 2;
    }
    DrainType type = 2;
    repeated string processors = 3;
}

message Sample {
    string name = 1;
    MetaData metadata = 2;
    repeated Source sources = 10;
    repeated Processor processors = 11;
    repeated Drain drains = 12;
}